File: MusDB/MusDB/Checker.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

using System.IO;
using System.Collections;

using WaterLibrary.MySQL;

using MySql.Data.MySqlClient;
using MySql;
using MusDB;
using WaterLibrary.Util;
using System.Security.Cryptography;

namespace MusDB
{
    class Checker
    {
        /// <summary>
        /// 私有构造
        /// </summary>
        private Checker() { }

        /// <summary>
        /// 文件转SHA256签名
        /// </summary>
        /// <param name="path"></param>
        /// <returns></returns>
        public static string ToSHA256(string path)
        {
            using SHA256 SHA256 = SHA256.Create();
            using FileStream File = new FileStream(path, FileMode.Open, FileAccess.Read);

            string result = BitConverter.ToString(SHA256.ComputeHash(File));

            return result;
        }

        /// <summary>
        /// 检查某路径下的所有文件并计数
        /// </summary>
        /// <param name="Path"></param>
        /// <param name="Count"></param>
        /// <returns>所有文件的数据列表</returns>
        public static List<(string Name, string MD5, string path, string file_type)> CheckFiles(string Path, ref (int flac, int mp3, int etc, int total) Count)
        {
            List<(string Name, string MD5, string path, string file_type)> Files = new();

            foreach (var el in new DirectoryInfo(Path).GetFileSystemInfos())
            {
                if (el is DirectoryInfo)
                {
                    Files.AddRange(CheckFiles(el.FullName, ref Count));
                    CLI.Line();
                }
                else
                {
                    CLI.Put(Count.total.ToString().PadLeft(4, '0'));
                    var temp = (FileInfo)el;
                    if (temp.Name.Contains(".flac"))
                    {
                        CLI.Line(" | flac  " + temp.Name);


                        var result = ToSHA256(temp.FullName);
                        Files.Add(new(temp.Name, result, temp.DirectoryName, "flac"));

                        CLI.InPosition(Console.WindowWidth / 5 * 3, Console.CursorTop - 1,
                           () => { CLI.Line(temp.DirectoryName); });

                        Count.flac++;
                    }
                    else if (temp.Name.Contains(".mp3"))
                    {
                        CLI.Line(" |  mp3  " + temp.Name);


                        var result = ToSHA256(temp.FullName);
                        Files.Add(new(temp.Name, result, temp.DirectoryName, "mp3"));

                        CLI.InPosition(Console.WindowWidth / 5 * 3, Console.CursorTop - 1,
                           () => { CLI.Line(temp.DirectoryName); });

                        Count.mp3++;
                    }
                    else
                    {
                        CLI.Line(temp.Name);
                        Files.Add(new(temp.Name, "", temp.DirectoryName, ""));

                        Count.etc++;
                    }
                    Count.total++;
                }

            }
            return Files;
        }

        public static IEnumerable<string> CheckETC(List<(string Name, string MD5, string path, string file_type)> Files)
        {
            return from el in Files where el.file_type == "" select el.Name;
        }

        public static List<List<string>> CheckConflict(List<(string Name, string MD5, string path, string file_type)> Files)
        {
            var conflicts = from el in (from el in Files group el by el.MD5) where el.Count() > 1 select el;
            List<List<string>> result = new();
            if (conflicts.Count() > 0)
            {
                foreach (var el in conflicts)
                {
                    List<string> it = new();
                    foreach (var inner_el in el)
                    {
                        it.Add(inner_el.Name);
                    }
                    result.Add(it);
                }
            }
            return result;
        }
    }
}

File: MusDB_fs/MusDB/Checker.fs
module App.Checker

open System
open System.IO
open System.Linq
open System.Collections.Generic
open System.Security.Cryptography
open App
open Util
open Mod
open CLI


type Checker =
    static member ToSha256 path =
        let file = new FileStream(path, FileMode.Open, FileAccess.Read)
        SHA256.Create() |> fun it -> it.ComputeHash file |> BitConverter.ToString

    static member GetFileSystemInfosList path = map id [ for x in (new DirectoryInfo(path)).GetFileSystemInfos() do x ]

    static member GetAllFiles (list:FileSystemInfo list) =
        match list with
        | x :: xs -> 
            match x with
            | :? DirectoryInfo as di ->
                Checker.GetAllFiles(Checker.GetFileSystemInfosList di.FullName)
            | :? FileInfo      as fi -> 
                match true with
                | _ when fi.Name.Contains ".flac" ->
                    { Name = fi.Name
                      Sha256 = Checker.ToSha256(fi.FullName)
                      Path = fi.DirectoryName
                      Type = "flac" } :: Checker.GetAllFiles xs
                | _ when fi.Name.Contains ".mp3"  ->
                    { Name = fi.Name
                      Sha256 = Checker.ToSha256(fi.FullName)
                      Path = fi.DirectoryName
                      Type = "mp3" } :: Checker.GetAllFiles xs
                | _  ->
                    { Name = fi.Name
                      Sha256 = Checker.ToSha256(fi.FullName)
                      Path = fi.DirectoryName
                      Type = "" } :: Checker.GetAllFiles xs
            | _  ->[]
        | [] -> []

    static member GetMusicFiles files = filter (fun x -> x.Type <> "") files
    static member GetOtherFiles files = filter (fun x -> x.Type = "") files

    static member CheckConflicts(files: File list) =
        [ for el in files.GroupBy(fun el -> el.Sha256) do
              if el.Count() > 1 then for it in el -> it ]
